ts/metrics/metric_cache_abstract.py

```diff
@@ -0,0 +1,462 @@
+           type for defining different operations, defaulted to counter metric type for Counter metrics
+        """
+        unit = "count"
+        req_id = self._get_req(idx)
+        self._add_or_update(
+            name=name,
+            value=value,
+            req_id=req_id,
+            unit=unit,
+            metric_type=metric_type,
+            dimensions=dimensions,
+        )
+
+    def add_time(
+        self,
+        name: str,
+        value: int or float,
+        idx=None,
+        unit: str = "ms",
+        dimensions: list = None,
+        metric_type: MetricTypes = MetricTypes.gauge,
+    ):
+        """
+        Add a time based metric like latency, default unit is 'ms'
+            Default metric type is gauge
+
+        Parameters
+        ----------
+        name : str
+            metric name
+        value: int
+            value of metric
+        idx: int
+            request_id index in batch
+        unit: str
+            unit of metric,  default here is ms, s is also accepted
+        dimensions: list
+            list of dimensions for the metric
+        metric_type: MetricTypes
+           type for defining different operations, defaulted to gauge metric type for Time metrics
+        """
+        if unit not in ["ms", "s"]:
+            raise merrors.MetricsCacheValueError(
+                "the unit for a timed metric should be one of ['ms', 's']"
+            )
+        req_id = self._get_req(idx)
+        self._add_or_update(
+            name=name,
+            value=value,
+            req_id=req_id,
+            unit=unit,
+            metric_type=metric_type,
+            dimensions=dimensions,
+        )
+
+    def add_size(
+        self,
+        name: str,
+        value: int or float,
+        idx=None,
+        unit: str = "MB",
+        dimensions: list = None,
+        metric_type: MetricTypes = MetricTypes.gauge,
+    ):
+        """
+        Add a size based metric
+            Default metric type is gauge
+
+        Parameters
+        ----------
+        name : str
+            metric name
+        value: int, float
+            value of metric
+        idx: int
+            request_id index in batch
+        unit: str
+            unit of metric, default here is 'MB', 'kB', 'GB' also supported
+        dimensions: list
+            list of dimensions for the metric
+        metric_type: MetricTypes
+           type for defining different operations, defaulted to gauge metric type for Size metrics
+        """
+        if unit not in ["MB", "kB", "GB", "B"]:
+            raise ValueError(
+                "The unit for size based metric is one of ['MB','kB', 'GB', 'B']"
+            )
+        req_id = self._get_req(idx)
+        self._add_or_update(
+            name=name,
+            value=value,
+            req_id=req_id,
+            unit=unit,
+            metric_type=metric_type,
+            dimensions=dimensions,
+        )
+
+    def add_percent(
```

Comment:
I wonder if instead we should have

```python
def add_counter():
def add_gauge():
def add_histogram():
```

Then on top of those we can add convenience wrappers like

```python
def add_percentage():
  add_gauge()
```